lib/src/rules/avoid_private_typedef_functions.dart

```diff
@@ -60,7 +60,14 @@ class Visitor extends SimpleAstVisitor {
 
   _countAndReport(String name, AstNode node) {
     final visitor = new _CountVisitor(name);
-    getCompilationUnit(node).root.visitChildren(visitor);
+    final units = getCompilationUnit(node)
+        .element
+        .library
+        .units
+        .map((cu) => cu.computeNode());
```

Comment:
I'm concerned about this. The method `computeNode` is generally quite expensive, so invoking it for every `typedef` seems likely to make this lint very slow. We should do some performance testing to determine whether that concern is valid.

Reply:
https://travis-ci.org/dart-lang/linter/jobs/355961824#L556

Reply:
I was just going to link to that!  For posterity, here's what it tells us:

```
---------------------------------------------------
Timings                                          ms
---------------------------------------------------
invariant_booleans                              444
public_member_api_docs                          257
avoid_private_typedef_functions                 142
```

It'd be interesting to disable this check and do another run for comparison?

Reply:
What code base(s) does that run the lint rules over, and how many parts do those code bases make use of?

Reply:
Itself.  We should get more representative for sure.  Moreover we have no parts so not sure what this tells us. :(

Reply:
I just ran it on flutter:

```
$ ./bin/cache/dart-sdk/bin/dart ~/dart/dart-lang/linter/bin/linter.dart --benchmark -q -c analysis_options_repo.yaml packages/flutter

---------------------------------------------------
Timings                                          ms
---------------------------------------------------
avoid_private_typedef_functions                1283
invariant_booleans                             1249
public_member_api_docs                          903
prefer_final_fields                             454
unnecessary_this                                236
package_api_docs                                208
avoid_renaming_method_parameters                206
cancel_subscriptions                            190
```

I think I can make an optim if there is only one unit to avoid `computeNode`. Stay tuned.

Reply:
A couple of additional optimizations:
- It could cache the list of units the first time they are computed and re-use it.
- It could avoid `computeNode` for the current compilation unit, given that it's the starting point for computing the list.

To be representative, the code base used for measurement also needs to have a representative number of `typedef` declarations.

Reply:
humm, after my following optim:

```diff
--- a/lib/src/rules/avoid_private_typedef_functions.dart
+++ b/lib/src/rules/avoid_private_typedef_functions.dart
@@ -60,11 +60,12 @@ class Visitor extends SimpleAstVisitor {

   _countAndReport(String name, AstNode node) {
     final visitor = new _CountVisitor(name);
-    final units = getCompilationUnit(node)
-        .element
-        .library
-        .units
-        .map((cu) => cu.computeNode());
+    final cu = getCompilationUnit(node);
+    final unitElements = cu.element.library.units;
+    print(unitElements.length);
+    final units = unitElements.length == 1
+        ? [cu]
+        : unitElements.map((e) => e.computeNode());
     for (final unit in units) {
       unit.visitChildren(visitor);
     }
```

it looks like `computeNode` is not really a problem (I see the same order of timing).

I guess the time is spent by `_CountVisitor`.

WDYT ?

Reply:
I can't tell from the diff where the time is being spent. I'd recommend profiling the code (either in observatory or by using some stopwatches to capture how long each piece takes).

Reply:
I added some stopwatches and they confirm the time is mainly spent in `_CountVisitor`.

```dart
  _countAndReport(String name, AstNode node) {
    final visitor = new _CountVisitor(name);
    final cu = getCompilationUnit(node);
    final unitElements = cu.element.library.units;
    print('--');
    print('nb of units: ${unitElements.length}');
    final sw1 = new Stopwatch()..start();
    final units = unitElements.length == 1
        ? [cu]
        : unitElements.map((e) => e.computeNode());
    print('create unit list: ${sw1.elapsedMicroseconds}');
    final sw2 = new Stopwatch()..start();
    for (final unit in units) {
      unit.visitChildren(visitor);
    }
    print('visit units: ${sw2.elapsedMicroseconds}');
    if (visitor.count <= 1) {
      rule.reportLint(node);
    }
  }
```

run on flutter:

```
$ ./bin/cache/dart-sdk/bin/dart ~/dart/dart-lang/linter/bin/linter.dart -q -c analysis_options_repo.yaml packages/flutter
--
nb of units: 1
create unit list: 22
visit units: 753
--
nb of units: 1
create unit list: 21
visit units: 423
--
nb of units: 1
create unit list: 1
visit units: 450
--
nb of units: 17
create unit list: 8
visit units: 593106
--
nb of units: 1
create unit list: 1
visit units: 29
--
nb of units: 1
create unit list: 1
visit units: 95
--
nb of units: 1
create unit list: 22
visit units: 501
--
nb of units: 17
create unit list: 1
visit units: 640530
817 files analyzed, 141183 issues found, in 19555 ms.
```

So I think we can go ahead with the current version in PR.

Reply:
Wow! I'm surprised that `computeNode` didn't take more time, but I'm also curious to know why it took an average of ~35000 ms / file to analyze the code in the libraries with 17 units.

But I can sign off on using `computeNode` :-).

Reply:
>why it took an average of ~35000 ms / file

microseconds not milliseconds ;)

Reply:
This is great.  Thanks for the follow-ups!